design-proposals/2022-10-27-rendering-modularization.md

```diff
@@ -0,0 +1,495 @@
+
+Which is to say that Drawable is a good concept which can encapsulate a lot of complexity.  So much so that it eventually encapsulates ALL the complexity, at least at that level of rendering.  It’s a good interface between the things that construct geometry and the things that render geometry.
+
+To keep the Drawable somewhat manageable, we added the concept of a Drawable Builder.  This is an object that you throw geometry at in a somewhat disordered way and it emits Drawables when you’re done.  MapLibre's [Buckets](https://github.com/maplibre/maplibre-gl-native/blob/main/src/mbgl/renderer/bucket.hpp) are similar, but not quite the same.  Perhaps they'll be adaptable.  We'll see.
+
+Drawables and Builders have a subclass for each supported rendering SDK.  One mistake we won’t bring over is using multiple inheritance for that.  You’re welcome.  
+
+It’s kind of obvious why you’d have a subclass of Drawable for each SDK.  You want to let the SDK represent data the way it wants (interleaved or not, 16 bit or 32 bit floats, all sorts of things), but it’s less obvious why the Builders need one subclass per SDK.
+
+Builders are actually doing the most specific SDK related work.  For OpenGL you build up fairly simple arrays of data, which MapLibre has good support for.  But for Metal, you can (and should) allocate space off a shared heap and put the data in there.  Each of these ends up looking fairly different and the way you wire things up at the low level is very different.  So Builders are very specific and let us hide all that.
+
+![Layer + Builder + Drawables](resources/fig2.png)
+
+Builders will be used primarily by Layers to emit Drawables for rendering.
+
+#### The way it is now:
+
+This is a big topic because it gets to the very heart of the critical changes we want to make to the toolkit.  There are a couple of different categories that are revealing and we'll use some examples.
+
+Most commonly there are *render layers*.  These are individual classes that know how to render, for instance, the [fill layer](https://github.com/maplibre/maplibre-gl-native/blob/main/src/mbgl/renderer/layers/render_fill_layer.cpp).  You'll see a bunch of these that start with "render" and end with "layer".
+
+What's going on here is that the toolkit has broken out a bunch of features from a tile, consolidated them together in a Bucket and then handed them off to a Layer to render.  Sure enough, if you look in the render_*_layer you can see it doing just that.
+
+Now it's not redoing the work of consolidation each frame, but it is doing a lot and a lot that is very specific to what the layer wants.  This leads to a few interesting issues.
+- That's a lot of commands that are fairly similar in every render layer.  Not a tiny amount of duplication there.
+- This is more specific logic than I want to see in a rendering loop.  Remember, this runs every frame.
```

Comment:
Writing in first person is fine I think, just make sure to state at the beginning somewhere the name of the author...

Reply:
I've modified the language.